### PR TITLE
Feature/add get display media

### PR DESCRIPTION
**Pull request by tmoldovan8x8**

### CODE REVIEW

File: MediaStreamTrack.js [hunk 1]
```diff
@@ -87,7 +87,6 @@ class MediaStreamTrack extends EventTarget(MEDIA_STREAM_TRACK_EVENTS) {
   }
 
   applyConstraints() {
-    throw new Error('Not implemented.');
```
Comment (saghul) [inline MediaStreamTrack.js:90]: Please revert this change.

File: android/src/main/java/com/oney/WebRTCModule/AbstractVideoCaptureController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
+package com.oney.WebRTCModule;
+
+import org.webrtc.VideoCapturer;
+
+public abstract class AbstractVideoCaptureController {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/AbstractVideoCaptureController.java:5]: Very nice!

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 3]
```diff
@@ -59,6 +77,24 @@
...
+
+        reactContext.addActivityEventListener(new BaseActivityEventListener() {
+            @RequiresApi(api = Build.VERSION_CODES.N)
+            @Override
+            public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
+                super.onActivityResult(activity, requestCode, resultCode, data);
+                if (requestCode == PERMISSION_REQUEST_CODE) {
+                    if (resultCode != Activity.RESULT_OK) {
+                        promise.reject("DOMException", "NotAllowedError");
+                        promise = null;
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:89]: Is it necessary to nullify the promise here?

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 4]
```diff
@@ -159,10 +159,11 @@ MediaStreamTrack getTrack(String id) {
      * if audio permission was not granted, there will be no "audio" key in
      * the constraints map.
      */
+    @RequiresApi(api = Build.VERSION_CODES.N)
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:162]: Why is this necessary? We currently support (and want to keep doing so) M.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 5]
```diff
@@ -241,9 +227,137 @@ void disposeTrack(String id) {
...
+            return;
+        }
+
+        Activity currentActivity = this.reactContext.getCurrentActivity();
+        if (currentActivity == null) {
+            promise.reject(new RuntimeException("No current Activity."));
+            return;
+        }
+
+        this.promise = promise;
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:248]: Can we call this something less generic than "promise" so we know it's related to getDisplayMedia please?

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 6]
```diff
@@ -241,9 +227,137 @@ void disposeTrack(String id) {
...
+
+        MediaProjectionManager mediaProjectionManager =
+            (MediaProjectionManager) currentActivity.getApplication().getSystemService(
+                Context.MEDIA_PROJECTION_SERVICE);
+        currentActivity.startActivityForResult(
+            mediaProjectionManager.createScreenCaptureIntent(), PERMISSION_REQUEST_CODE);
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.N)
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:257]: What about M? O even L? We should return an error at least.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 7]
```diff
@@ -241,9 +227,137 @@ void disposeTrack(String id) {
...
+
+        successCallback.accept(streamId, tracksInfo);
+    }
+
+    private VideoTrack createScreenTrack() {
+        DisplayMetrics displayMetrics = getDisplayMetrics();
+        int width = displayMetrics.widthPixels;
+        int height = displayMetrics.heightPixels;
+        int fps = 30;
+        ScreenVideoCaptureController videoCaptureController = new ScreenVideoCaptureController(width, height, fps, mediaProjectionPermissionResultData);
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:316]: ScreenCaptureController?

File: android/src/main/java/com/oney/WebRTCModule/ScreenVideoCaptureController.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package com.oney.WebRTCModule;
+
+import android.content.Intent;
+import android.media.projection.MediaProjection;
+import android.util.Log;
+
+import org.webrtc.ScreenCapturerAndroid;
+import org.webrtc.VideoCapturer;
+
+public class ScreenVideoCaptureController extends AbstractVideoCaptureController {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/ScreenVideoCaptureController.java:10]: ScreenCaptureController?

File: android/src/main/java/com/oney/WebRTCModule/ScreenVideoCaptureController.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    protected VideoCapturer createVideoCapturer() {
+        VideoCapturer videoCapturer = new ScreenCapturerAndroid(
+            mediaProjectionPermissionResultData,
+            new MediaProjection.Callback() {
+                @Override
+                public void onStop() {
+                    Log.w(TAG, "User revoked permission to capture the screen.");
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/ScreenVideoCaptureController.java:35]: How can we better handle this error?

File: getDisplayMedia.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import { Platform, NativeModules } from 'react-native';
+
+import MediaStream from './MediaStream';
+import MediaStreamError from './MediaStreamError';
+
+const { WebRTCModule } = NativeModules;
+
+export default function getDisplayMedia(constraints) {
+    if (Platform.OS !== 'android') {
+        return Promise.reject(new TypeError());
```
Comment (saghul) [inline getDisplayMedia.js:12]:
```suggestion
        return Promise.reject(new Error('Unsupported platform'));
```

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL!

File: android/src/main/java/com/oney/WebRTCModule/ScreenVideoCaptureController.java (hunk 9 above)
Comment (tmoldovan8x8) [reply android/src/main/java/com/oney/WebRTCModule/ScreenVideoCaptureController.java:35]: Actually, this is called when the MediaProjection is stopped, I've changed the logging message. From what I can tell, this can only be stopped by manually calling the stop method on it.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java (hunk 3 above)
Comment (tmoldovan8x8) [reply android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:89]: It's necessary for the logic in getDisplayMedia(Promise promise) to work, where we verify that there isn't another task in progress.

File: android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java [hunk 11]
```diff
@@ -241,11 +221,137 @@ void disposeTrack(String id) {
...
         TrackPrivate track = tracks.get(trackId);
-        if (track != null && track.videoCaptureController != null) {
-            track.videoCaptureController.switchCamera();
+        if (track != null && track.videoCaptureController instanceof CameraCaptureController) {
+            CameraCaptureController cameraCaptureController = (CameraCaptureController) track.videoCaptureController;
+            cameraCaptureController.switchCamera();
         }
     }
 
+    public void getDisplayMedia(Promise promise) {
```
Comment (saghul) [inline android/src/main/java/com/oney/WebRTCModule/GetUserMediaImpl.java:230]:
```suggestion
    void getDisplayMedia(Promise promise) {
```
